### ISSUE TITLE
ROW type hash implementation throws misleading message

### ISSUE THREAD

**Issue by dain** [prestodb/presto#5278]
Message should say `hash` is not supported for...

```
com.facebook.presto.spi.PrestoException: ROW comparison not supported for fields with null elements
    at com.facebook.presto.type.RowType.checkElementNotNull(RowType.java:216)
    at com.facebook.presto.type.RowType.hash(RowType.java:206)
```

**Comment by lacbs** [comment]
It seems that this was intentional: https://github.com/prestodb/presto/commit/43c6fa7ef6d8f6994f650df00628401bb5d44dee

Any reason why not compare row with null elements?

**Comment by geraint0923** [comment]
I think the problem might be that
```
row(1,2) = row(1,2)        => true
row(null,2) = row(1,2)     => null
row(null,2) = row(1,3)     => false
```
However for now `type.equalTo` only return `boolean` which is not a object and could not express `null`.

**Comment by lacbs** [comment]
imo, `row(null,2) = row(1,2)` should be `false`, not null. The row itself is not null.
`row(null,2) = row(1,2)` being `null` seems a weird behavior.
Complex one too, since `row(null,2) = row(1,3)` is `false`, why `false` precedes `null` anyway?

I agree in `row(...) = null` returning null (like `array[1] = CAST(null as array(bigint))` does).

btw, this same problem also happens to array: `select array[null, 1] = array[null, 1];` throws an exception: `ARRAY comparison not supported for arrays with null elements`.
Should we treat this in this issue or https://github.com/prestodb/presto/issues/3569 or a new one?

**Comment by martint** [comment]
Row comparison semantics are equivalent to doing a pairwise comparison of the elements and joining them with AND. So:

Row(a, b) = Row(c, d) 

Is equivalent to:

a = c AND b = d

Then, standard semantics apply for how nulls are treated:

null AND true => null
null AND false => false

The way to think about it is "null means the actual value is unknown". Given that notion, if any pair of fields don't match, we know for sure that the equality is false. If all pairs of fields but one match, as the latter is a comparison with an "unknown" value, we can't tell whether the equality is true or false, so it returns "unknown".

**Comment by lacbs** [comment]
Got it, that's a really good explanation, thanks a lot :)

**Comment by haozhun** [comment]
Fixed by #6429.
